<a href="https://colab.research.google.com/github/priyasree17/PersonalPortfolio/blob/main/EnergyConsumption.ipynb" target="_parent"><img src="https://colab.research.google.com/assets/colab-badge.svg" alt="Open In Colab"/></a>

# **ENERGY CONSUMPTION AND PREDICTION - DATA EXPLORATION**


The dataset contain time-series data about household energy consumption, where measurements of power, voltage, and other electrical readings are collected over time. The data covers 377,022 entries (rows) with 9 columns.



**DATASET**



**Date :** The date on which the measurements were recorded.

**Time :** The time of day when the measurement was taken.

**Global_active_power :** The total active power consumed across the household.

**Global_reactive_power :**	The total reactive power consumed across the household.

**Voltage :** The voltage level in the household, measured in volts (V).

**Global_intensity :** The intensity of current drawn by the appliances in the household.

**Sub_metering_1 :** Energy sub-metering for the first area in the household.

**Sub_metering_2 :**  Energy sub-metering for the second area in the household.

**Sub_metering_2 :** Energy sub-metering for the third area in the household.

# **1. Defining Problem Statement and Analyzing Basic Metrics**


The main objective of this project is to analyze the dataset "Energy consumption" and to  predict it.

# **2. Import library and Load the dataset**

In [40]:
import pandas as pd

In [2]:
!gdown --fuzzy https://drive.google.com/file/d/1bvaXJJqNObOCkX-i475BNxpidk024pyx/view

Downloading...
From (original): https://drive.google.com/uc?id=1bvaXJJqNObOCkX-i475BNxpidk024pyx
From (redirected): https://drive.google.com/uc?id=1bvaXJJqNObOCkX-i475BNxpidk024pyx&confirm=t&uuid=cdde991a-1426-4d71-aedc-8f907dc1a49e
To: /content/household_power_consumption.txt
100% 133M/133M [00:01<00:00, 120MB/s]


In [3]:
df = pd.read_csv('/content/household_power_consumption (2).txt',sep=";")

In [4]:
# Describing the statistical summary of numerical type data
df.describe()

Sub_metering_3
count   373093.000000
mean         5.548204
std          8.073037
min          0.000000
25%          0.000000
50%          0.000000
75%         17.000000
max         20.000000

# **OBSERVATIONS:**


**count :** 373,093 non-null entries out of 377,022 total rows. This means there are 3,929 missing values in the Sub_metering_3 column.

**Mean (Average):**
 The average value of Sub_metering_3 is 5.55 watt-hours, indicating that on average, the energy consumption for this sub-metered area is relatively low.

**Standard Deviation (std):**

The standard deviation is 8.07, which suggests there is considerable variability in the energy consumption for this area. Some values are much higher or lower than the average.

**Minimum (min):**

The minimum value is 0.00, meaning there are periods with no energy consumption in this sub-metered area.

**25th Percentile (25%):**

The 25th percentile is 0.00, indicating that at least 25% of the data points show no energy consumption for this area.

**50th Percentile (Median or 50%):**

The median value is also 0.00, meaning that at least 50% of the data points have zero energy consumption in this sub-metered area. This suggests that for half of the recorded observations, the area being measured by Sub_metering_3 did not consume any energy.

**75th Percentile (75%):**

The 75th percentile is 17.00, meaning that 75% of the data points have energy consumption less than or equal to 17 watt-hours. This indicates that higher values for energy consumption are less frequent but occur for about 25% of the data.

**Maximum (max):**

The maximum value recorded is 20.00 watt-hours, showing that the highest energy consumption recorded for this sub-metered area is 20 watt-hours.

In [5]:

df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 377022 entries, 0 to 377021
Data columns (total 9 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Date                   377022 non-null  object 
 1   Time                   377021 non-null  object 
 2   Global_active_power    377021 non-null  object 
 3   Global_reactive_power  377021 non-null  object 
 4   Voltage                377021 non-null  object 
 5   Global_intensity       377021 non-null  object 
 6   Sub_metering_1         377021 non-null  object 
 7   Sub_metering_2         377021 non-null  object 
 8   Sub_metering_3         373093 non-null  float64
dtypes: float64(1), object(8)
memory usage: 25.9+ MB


# **OBSERVATIONS:**


*The dataset consists of 377,022 rows and 9 columns.

*Most columns, including "Date," "Time," "Global_active_power," "Global_reactive_power," "Voltage," "Global_intensity," "Sub_metering_1," and "Sub_metering_2," are stored as strings (object datatype), indicating that they contain textual or numeric data that hasn't been converted to numerical types yet.

*The only column with a float datatype is "Sub_metering_3," which contains floating-point numbers representing energy sub-metering data.



In [6]:
# Statistical summary of categorical type data
df.describe(include = object)

Date      Time Global_active_power Global_reactive_power Voltage  \
count      377022    377021              377021                377021  377021   
unique        264      1440                3679                   436    2522   
top     27/4/2007  17:24:00                   ?                 0.000       ?   
freq         1440       262                3928                 95089    3928   

       Global_intensity Sub_metering_1 Sub_metering_2  
count            377021         377021         377021  
unique              205             79             80  
top               1.000          0.000          0.000  
freq              40128         341651         258818

In [8]:
# Dimension of the dataframe
df.shape

(377022, 9)

In [14]:
#Unique Values in Each Column
df.nunique()

Date                      264
Time                     1440
Global_active_power      3679
Global_reactive_power     436
Voltage                  2522
Global_intensity          205
Sub_metering_1             79
Sub_metering_2             80
Sub_metering_3             21
dtype: int64

# **3. Check for missing values**

This is both a data cleaning and data preprocessing step. Identifying and handling missing values is considered data cleaning since it involves addressing the issue of incomplete data. Depending on the extent of missing data, you may need to decide how to handle it, either by imputing values or removing the affected rows/columns. Additionally, it is also a data preprocessing step since having missing values can impact the effectiveness of subsequent analyses, and addressing them helps ensure the data is in a suitable form for analysis.

In [15]:
# Display the count of missing values for each column
df.isnull().sum()

Date                        0
Time                        1
Global_active_power         1
Global_reactive_power       1
Voltage                     1
Global_intensity            1
Sub_metering_1              1
Sub_metering_2              1
Sub_metering_3           3929
dtype: int64

In [17]:
# Calculate the missing values percentage for each column and round to three decimal places
round((df.isnull().sum()/len(df))*100,3)

Date                     0.000
Time                     0.000
Global_active_power      0.000
Global_reactive_power    0.000
Voltage                  0.000
Global_intensity         0.000
Sub_metering_1           0.000
Sub_metering_2           0.000
Sub_metering_3           1.042
dtype: float64

In [19]:
#data cleaning---->Drop rows with null values
df_cleaned = df.dropna()
df_cleaned

Date      Time Global_active_power Global_reactive_power  \
0       16/12/2006  17:24:00               4.216                 0.418   
1       16/12/2006  17:25:00               5.360                 0.436   
2       16/12/2006  17:26:00               5.374                 0.498   
3       16/12/2006  17:27:00               5.388                 0.502   
4       16/12/2006  17:28:00               3.666                 0.528   
...            ...       ...                 ...                   ...   
377016    4/9/2007  13:00:00               1.496                 0.160   
377017    4/9/2007  13:01:00               1.482                 0.136   
377018    4/9/2007  13:02:00               1.476                 0.114   
377019    4/9/2007  13:03:00               1.480                 0.114   
377020    4/9/2007  13:04:00               1.476                 0.114   

        Voltage Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
0       234.840           18.400          0.000          1.000            17.0  
1       233.630           23.000          0.000          1.000            16.0  
2       233.290           23.000          0.000          2.000            17.0  
3       233.740           23.000          0.000          1.000            17.0  
4       235.680           15.800          0.000          1.000            17.0  
...         ...              ...            ...            ...             ...  
377016  241.970            6.200          0.000          0.000            18.0  
377017  241.680            6.200          0.000          0.000            18.0  
377018  241.950            6.200          0.000          0.000            18.0  
377019  242.020            6.200          0.000          0.000            18.0  
377020  242.020            6.200          0.000          0.000            18.0  

[373093 rows x 9 columns]

In [24]:
#Fill missing values
df_filled = df.fillna(df['Sub_metering_3'].mean())
df_filled

Date      Time Global_active_power Global_reactive_power  \
0       16/12/2006  17:24:00               4.216                 0.418   
1       16/12/2006  17:25:00               5.360                 0.436   
2       16/12/2006  17:26:00               5.374                 0.498   
3       16/12/2006  17:27:00               5.388                 0.502   
4       16/12/2006  17:28:00               3.666                 0.528   
...            ...       ...                 ...                   ...   
377017    4/9/2007  13:01:00               1.482                 0.136   
377018    4/9/2007  13:02:00               1.476                 0.114   
377019    4/9/2007  13:03:00               1.480                 0.114   
377020    4/9/2007  13:04:00               1.476                 0.114   
377021           4  5.548204            5.548204              5.548204   

         Voltage Global_intensity Sub_metering_1 Sub_metering_2  \
0        234.840           18.400          0.000          1.000   
1        233.630           23.000          0.000          1.000   
2        233.290           23.000          0.000          2.000   
3        233.740           23.000          0.000          1.000   
4        235.680           15.800          0.000          1.000   
...          ...              ...            ...            ...   
377017   241.680            6.200          0.000          0.000   
377018   241.950            6.200          0.000          0.000   
377019   242.020            6.200          0.000          0.000   
377020   242.020            6.200          0.000          0.000   
377021  5.548204         5.548204       5.548204       5.548204   

        Sub_metering_3  
0            17.000000  
1            16.000000  
2            17.000000  
3            17.000000  
4            17.000000  
...                ...  
377017       18.000000  
377018       18.000000  
377019       18.000000  
377020       18.000000  
377021        5.548204  

[377022 rows x 9 columns]

In [26]:
df_filled = df.fillna(0)
df_filled

Date      Time Global_active_power Global_reactive_power  \
0       16/12/2006  17:24:00               4.216                 0.418   
1       16/12/2006  17:25:00               5.360                 0.436   
2       16/12/2006  17:26:00               5.374                 0.498   
3       16/12/2006  17:27:00               5.388                 0.502   
4       16/12/2006  17:28:00               3.666                 0.528   
...            ...       ...                 ...                   ...   
377017    4/9/2007  13:01:00               1.482                 0.136   
377018    4/9/2007  13:02:00               1.476                 0.114   
377019    4/9/2007  13:03:00               1.480                 0.114   
377020    4/9/2007  13:04:00               1.476                 0.114   
377021           4         0                   0                     0   

        Voltage Global_intensity Sub_metering_1 Sub_metering_2  Sub_metering_3  
0       234.840           18.400          0.000          1.000            17.0  
1       233.630           23.000          0.000          1.000            16.0  
2       233.290           23.000          0.000          2.000            17.0  
3       233.740           23.000          0.000          1.000            17.0  
4       235.680           15.800          0.000          1.000            17.0  
...         ...              ...            ...            ...             ...  
377017  241.680            6.200          0.000          0.000            18.0  
377018  241.950            6.200          0.000          0.000            18.0  
377019  242.020            6.200          0.000          0.000            18.0  
377020  242.020            6.200          0.000          0.000            18.0  
377021        0                0              0              0             0.0  

[377022 rows x 9 columns]

In [27]:
#converting object type into float type
df['Global_active_power'] = pd.to_numeric(df['Global_active_power'],errors = 'coerce')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 377022 entries, 0 to 377021
Data columns (total 9 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Date                   377022 non-null  object 
 1   Time                   377021 non-null  object 
 2   Global_active_power    373093 non-null  float64
 3   Global_reactive_power  377021 non-null  object 
 4   Voltage                377021 non-null  object 
 5   Global_intensity       377021 non-null  object 
 6   Sub_metering_1         377021 non-null  object 
 7   Sub_metering_2         377021 non-null  object 
 8   Sub_metering_3         373093 non-null  float64
dtypes: float64(2), object(7)
memory usage: 25.9+ MB


In [30]:
df['Global_reactive_power'] = pd.to_numeric(df['Global_reactive_power'],errors = 'coerce')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 377022 entries, 0 to 377021
Data columns (total 9 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Date                   377022 non-null  object 
 1   Time                   377021 non-null  object 
 2   Global_active_power    373093 non-null  float64
 3   Global_reactive_power  373093 non-null  float64
 4   Voltage                377021 non-null  object 
 5   Global_intensity       377021 non-null  object 
 6   Sub_metering_1         377021 non-null  object 
 7   Sub_metering_2         377021 non-null  object 
 8   Sub_metering_3         373093 non-null  float64
dtypes: float64(3), object(6)
memory usage: 25.9+ MB


In [31]:
df['Voltage'] = pd.to_numeric(df['Voltage'],errors = 'coerce')
df['Global_intensity'] = pd.to_numeric(df['Global_intensity'],errors = 'coerce')
df['Sub_metering_1'] = pd.to_numeric(df['Sub_metering_1'],errors = 'coerce')
df['Sub_metering_2'] = pd.to_numeric(df['Sub_metering_2'],errors = 'coerce')
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 377022 entries, 0 to 377021
Data columns (total 9 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Date                   377022 non-null  object 
 1   Time                   377021 non-null  object 
 2   Global_active_power    373093 non-null  float64
 3   Global_reactive_power  373093 non-null  float64
 4   Voltage                373093 non-null  float64
 5   Global_intensity       373093 non-null  float64
 6   Sub_metering_1         373093 non-null  float64
 7   Sub_metering_2         373093 non-null  float64
 8   Sub_metering_3         373093 non-null  float64
dtypes: float64(7), object(2)
memory usage: 25.9+ MB


In [44]:
df.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 377022 entries, 0 to 377021
Data columns (total 9 columns):
 #   Column                 Non-Null Count   Dtype  
---  ------                 --------------   -----  
 0   Date                   1 non-null       float64
 1   Time                   0 non-null       float64
 2   Global_active_power    373093 non-null  float64
 3   Global_reactive_power  373093 non-null  float64
 4   Voltage                373093 non-null  float64
 5   Global_intensity       373093 non-null  float64
 6   Sub_metering_1         373093 non-null  float64
 7   Sub_metering_2         373093 non-null  float64
 8   Sub_metering_3         373093 non-null  float64
dtypes: float64(9)
memory usage: 25.9 MB


**OBSERVATION :**

The dataset contains 377,022 rows and 9 columns, with most columns having 373,093 non-null entries. The "Date" column has only 1 non-null entry, and the "Time" column is completely missing data (0 non-null entries). All columns are of float64 datatype.

In [43]:
df.describe()

Date  Time  Global_active_power  Global_reactive_power        Voltage  \
count   1.0   0.0        373093.000000          373093.000000  373093.000000   
mean    4.0   NaN             1.098677               0.123785     239.012341   
std     NaN   NaN             1.160628               0.110913       3.669986   
min     4.0   NaN             0.082000               0.000000     223.490000   
25%     4.0   NaN             0.266000               0.000000     236.330000   
50%     4.0   NaN             0.472000               0.106000     239.410000   
75%     4.0   NaN             1.526000               0.194000     241.660000   
max     4.0   NaN            10.670000               1.148000     251.700000   

       Global_intensity  Sub_metering_1  Sub_metering_2  Sub_metering_3  
count     373093.000000   373093.000000   373093.000000   373093.000000  
mean           4.704385        1.226992        1.576741        5.548204  
std            4.900087        6.446965        6.468195        8.073037  
min            0.400000        0.000000        0.000000        0.000000  
25%            1.200000        0.000000        0.000000        0.000000  
50%            2.200000        0.000000        0.000000        0.000000  
75%            6.400000        0.000000        1.000000       17.000000  
max           46.400000       78.000000       78.000000       20.000000

In [37]:
#dimensions of the DataFrame.
df.shape

(377022, 9)

In [39]:
#Unique Values in Each Column
df.nunique()

Date                        1
Time                        0
Global_active_power      3678
Global_reactive_power     435
Voltage                  2521
Global_intensity          204
Sub_metering_1             78
Sub_metering_2             79
Sub_metering_3             21
dtype: int64